src/lib/homestore/shard_manager.cpp

```diff
@@ -20,4 +124,74 @@ ShardManager::Result< InfoList > HSHomeObject::_list_shards(pg_id pg) const {
     return folly::makeUnexpected(ShardError::UNKNOWN_PG);
 }
 
+void HSHomeObject::on_shard_message_commit(int64_t lsn, sisl::blob const& header, sisl::blob const& key,
+                                           void* user_ctx) {
+    const ReplicationMessageHeader* msg_header = r_cast<const ReplicationMessageHeader*>(header.bytes);
+    folly::Promise<ShardManager::Result<ShardInfo> > *promise = nullptr;
+    // user_ctx will be nullptr when:
+    // 1. on the follower side
+    // 2. on the leader side but homeobject restarts and replay all commited log entries from the last commit id;
+    if (user_ctx != nullptr) {
+        promise = r_cast<folly::Promise<ShardManager::Result<ShardInfo> >*> (user_ctx);      
+    }
+
+    if (msg_header->header_crc != msg_header->calculate_crc()) {
+        LOGWARN("replication message header is corrupted with crc error, lsn:{}", lsn);
+        if (promise) {
+            promise->setValue(folly::makeUnexpected(ShardError::INVALID_ARG));	  
+        }
+        return;
+    }
+
+    std::string string_value;
+    string_value.append(r_cast<char*>(header.bytes + sizeof(ReplicationMessageHeader)), msg_header->payload_size);
```

Comment:
should be
`string_value.append(r_cast<char*>(msg_header), sizeof(ReplicationMessageHeader))`? 

please correct me if i am wrong

Reply:
no, string_value is used to hold the msg body.